themes/bootstrap3/theme.config.php

```diff
@@ -11,14 +11,14 @@
         'flex-fallback.css::lt IE 10', // flex polyfill
     ],
     'js' => [
-        'vendor/jquery.min.js',
-        'vendor/bootstrap.min.js',
-        'vendor/bootstrap-accessibility.min.js',
-        'vendor/validator.min.js',
-        'lib/form-attr-polyfill.js', // input[form] polyfill, cannot load conditionally, since we need all versions of IE
-        'lib/autocomplete.js',
-        'common.js',
-        'lightbox.js',
+        ['file' => 'vendor/jquery.min.js', 'priority' => 100],
```

Comment:
As I suggested above, I don't think we necessarily need to give every file a unique priority value. I was viewing this as a way of grouping things with common priority levels, rather than a system for explicitly setting the exact order of everything. Thus, I'd group them together in as few levels as possible. If files don't depend on one another, they should get the same level.

Reply:
That's true for custom themes, but i think that VuFind's default themes should have priority values anyway. Only if this is the case, the custom themes can set a distinct priority which is e.g. exactly between jquery.min.js and bootstrap.min.js. If a custom theme chooses to add additional values, it may do so without setting any priority.

Also, the current implementation will add all priority-based entries at the top in the correct order, and append all non priority-based entries behind that. If the default themes' entries would have no priority, the custom themes' entries with a priority would be above that, which might lead to runtime errors.

Reply:
That makes sense, though I wonder if it would at least be helpful to establish groupings through the priority numbering (for example, what if all the vendor dependencies were 1xxx and all the local ones were 2xxx? But maybe I'm overthinking this).

Reply:
If we add the grouping as first number in the priority, this would also affect the sorting and no longer allow a local dependency to be sorted within vendor dependencies. On the other hand, if we would add an additional 'group' key, this would already be possible but only be for used better readability and not have any real effect inside the code.

Reply:
I think we may not be entirely understanding each other. :-) I'm not suggesting that the first digit of the priority number should have any special meaning within the code, or that we need to add a specific group key for different "types" of dependencies, or anything like that. I'm just suggesting that if we're choosing arbitrary integers for setting priority, we could come up with a convention for how the core code uses those numbers... so, for example, vendor dependencies could be 1010, 1020, 1030, etc. and VuFind dependencies could be 2010, 2020, 2030, etc. An end user could feel free to break this pattern if there was a reason to, and the code wouldn't care. I just thought it might be more readable if there was some kind of visible pattern in the built-in defaults.

Reply:
(And again, maybe this is not a good idea -- I certainly don't insist upon it; I just want to be sure I'm being clear about what my idea actually is!)

Reply:
Thanks for clarifying! I think the main question is this is well understandable by the users, because they need to keep in mind that e.g. 2020 would be higher ranked than 1030, followed e.g. by 40 (or 3040) and so on.

Maybe we should discuss this at the next community call to get some more opinions?

Reply:
Agreed. This is already on the planned agenda for the next community call, so we can definitely discuss in more detail at that time!

Reply:
@mtrojan-ub, thanks for the continued work on this! I wonder if it has become more complicated than I intended, though -- my original thought was that the priority values would be sorted as simple integers, but that we'd use a three-digit convention in the core code to improve readability, demonstrate some logical grouping of related files, and keep things unambiguous (since if we use the same number of digits for all values, numeric and alphabetic sort orders work out the same). It looks like you've taken steps to enforce three digit numbers all the time, and I'm not exactly sure what canonizePriority is for -- are you sorting on the second digit rather than the first? Anyway, I apologize if I expressed my ideas unclearly and led you to do more work than was necessary... but on the other hand, if you are happy with what you've come up with, I'm happy to discuss it further and make sure I understand it properly. :-)

Reply:
@demiankatz, Thanks for having another look! To answer your questions:
- canonizePriority is just a function to extract the sorting value from a priority. So, if we have a four-digit priority with the first digit being ignored for sorting, the function simply strips the first digit. Maybe another name might be better? (something like extractPrioritySortingValue, maybe you have a better idea?)
- I thought it might be better to restrict priorities to 3-digit-values to make things clearer (and you immediately get an error message if you chose the wrong scheme). We could also open up the scheme for all kinds of numbers, but what shall happen then? Will the first digit be ignored for sorting as well? What happens if a priority with only 1 digit is given / is it allowed and treated without a group, or should there be an error?

Reply:
I think the point I am confused about is why we are not using the first digit for sorting. My thought was that third-party vendor dependencies always need to load first -- that's why we give them 1xx numbers. Then locally-developed libraries load next, because they might depend on third-party code but they do not contain VuFind-specific logic -- that's why we give them 2xx numbers. Then very specific, VuFind-oriented code loads last -- that's why we give them 3xx numbers. My thought was that we'd just sort on the whole number, not ignore the first digit. And of course users would be free to change priorities as appropriate, but for core code, this convention seemed adequate. So I was expecting a numbering scheme like 100, 110, 120, 210, 220, 310, rather than 100, 110, 120, 230, 240, 350. Like I said, I'm certainly happy to discuss the merits of a different solution -- but I just want to be clear about what my original idea was, since I don't think we walked away from that conversation with exactly the same ideas. :-)

Reply:
Thanks for the explanation!

To make sure that i understand correctly this time: Do you mean that a user should be able to e.g. set one of his custom libraries to a priority of 115 if he wants to load it between third-party vendor dependencies with 110 and 120, even if his 115 is not a third-party itself?

Reply:
Yes, exactly! I intended this to be a convention/best practice but not an enforced requirement.

Reply:
OK, so now i re-simplified that. Please have another look!

Reply:
This looks good to me. I'm not seeing any Travis failures here; am I missing something?

In any case, at this point, is there any outstanding work here, or do you feel this is ready to merge? If you're happy with what we have, then I can give it one more thorough review, and merge it if I find no further concerns.

Thanks again for all of the work on this!

Reply:
The travis problems seem to be gone, and the version should be complete by now so it should be ready to merge!
Thank you again very much for your support!